### PR TITLE
[serial] Initial commit of some of the serial port handling bits.

### PR DESCRIPTION
**Pull request by brendank310**
Pull request intended for early review, not merging.

Signed-off-by: Brendan Kerrigan kerriganb@ainfosec.com

### CODE REVIEW

File: common/serial/Makefile [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+################################################################################
+# FLAGS
+################################################################################
+
+CROSS_CC=~/opt/cross/bin/x86_64-elf-gcc
+CROSS_CXX=~/opt/cross/bin/x86_64-elf-g++
+CROSS_LD=~/opt/cross/bin/x86_64-elf-ld
+
+# CC=$(CROSS_CC)
```
Comment (rianquinn) [inline common/serial/Makefile:30]: I would just get rid of these

File: common/serial/port_io.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#ifndef PORT_IO__H
```
Comment (brendank310) [inline common/serial/port_io.h:1]: The arch stuff got a little convoluted, and is probably mostly unnecessary

File: common/serial/Makefile [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+################################################################################
+
+TARGET_NAME=bf_serial
+TARGET_TYPE=lib
+TARGET_CROSS_COMPILED=false
+
+OBJDIR=.build
+OUTDIR=../bin
+
+SOURCES=serial.cpp arch/arch_linux.cpp
```
Comment (rianquinn) [inline common/serial/Makefile]: Just check your ".build" to make sure this is working properly. You should see a arch folder created that has your object file in it. If it's putting the object file in a strange place, you might need to use VPATH

File: common/serial/arch.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#include <bf_stdint.h>
+
+class arch_handler {
+
+public:
+	arch_handler() {}
+	virtual ~arch_handler() {}
+
+	virtual void port_write(uint16_t port, uint8_t value) = 0;
+	virtual void port_write(uint16_t port, uint16_t value) = 0;
```
Comment (rianquinn) [inline common/serial/arch.h:13]: since c++ sucks by the fact that it could do type conversions for you.... I would not use function overloading for this. that is..... make a port_write8 and port_write16

File: common/serial/arch.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+	virtual uint8_t port_read(uint16_t port) = 0;
```
Comment (rianquinn) [inline common/serial/arch.h:15]: port_read8?

File: common/serial/arch/arch_linux.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#ifndef ARCH_LINUX__H
+#define ARCH_LINUX__H
+
+#include <arch.h>
+#include <stdio.h>
+
+#define TRACE() printf("%s:%d\n", __PRETTY_FUNCTION__, __LINE__)
```
Comment (rianquinn) [inline common/serial/arch/arch_linux.h]: This should either be in the .cpp, or temporary left here until we can move it somewhere else.

File: common/serial/port_io.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+#define PORT_IO__H
+
+#include <arch.h>
+
+class port_io {
+
+public:
+	port_io(arch_handler *arch) : m_arch(arch)
```
Comment (rianquinn) [inline common/serial/port_io.h]: your going to need a third implementation of this. The third on would be your "stub" that would be used during unit testing

File: common/serial/serial.cpp [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include <serial.h>
+#include <arch_linux.h>
+
+struct __divisor_rate divisor_table[] = {
```
Comment (rianquinn) [inline common/serial/serial.cpp]: Would be great if you provided a spec for this

File: common/serial/serial.h [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+ *  http://www.sci.muni.cz/docs/pc/serport.txt
+ *
+ */
+
+#define COM1_IO_PORT 0x3f8
+#define COM2_IO_PORT 0x2f8
+#define COM3_IO_PORT 0x3e8
+#define COM4_IO_PORT 0x2e8
+
+/* Limits */
```
Comment (rianquinn) [inline common/serial/serial.h]: These comments should all be // since this is cpp

File: common/serial/serial.h [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+#define DATA_SIZE_MASK          0xFC
+
+struct __data_bits {
+    uint8_t data_size;
+    uint8_t data_bits;
+};
+
+extern struct __data_bits data_size_table[];
+
+/* Receive buffer register */
```
Comment (rianquinn) [inline common/serial/serial.h]: spec?

File: common/serial/serial.h [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    two registers.
+*/
+#define DLAB                    ((uint8_t)0x80)
+
+#define DATA_SIZE(x) (x-MIN_DATA_SIZE)
+
+class serial_port : public port_io {
+
+public:
+    serial_port(uint8_t port = 1, uint32_t baud = DEFAULT_BAUD_RATE, uint8_t data_bits = 8,
```
Comment (rianquinn) [inline common/serial/serial.h]: to help with unit testing, your going to want to provide the serial class with the port_io object that it plans to use. This way, during the unit test you can provide it with the stub, and in the real case, you can provide it with the with the linux version, or the real port_io commands.

File: common/serial/port_io.h [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	~port_io(void) 
+	{
+		delete m_arch;
+	}
+
+	void port_write(uint16_t port, uint8_t value)
+	{
+		m_arch->port_write(port, value);
```
Comment (rianquinn) [inline common/serial/port_io.h]: We will have to talk about how far we want to generalize this stuff. Not sure an arch class is needed. It might be.... not sure. In the end.... the actual port IO will have to come from code written in NASM as I don't want to use inline assembly.

File: common/serial/serial.h [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+public:
+    serial_port(uint8_t port = 1, uint32_t baud = DEFAULT_BAUD_RATE, uint8_t data_bits = 8, 
+                PARITY_MODE parity = NONE, uint8_t stop_bits = 1);
+    ~serial_port(void);
+
+    int32_t open(void);
+    int32_t close(void);
+
+    void set_baud_rate(uint32_t baud);
```
Comment (rianquinn) [inline common/serial/serial.h]: Most of these functions don't have any return values, which is going to make unit testing really difficult. It's also going to make it difficult to know if the functions are actually working.

File: common/serial/Makefile (hunk 1 above)
Comment (brendank310) [inline common/serial/Makefile:30]: Is the cross compilation handled in another way? Most of these Makefiles for serial will probably change significantly.

File: common/serial/Makefile (hunk 3 above)
Comment (brendank310) [inline common/serial/Makefile]: Nope that didn't fly for me. I got around it by manually touching the file. We should figure out what's necessary to support this scenario, as it's fairly common.

File: common/serial/arch.h (hunk 4 above)
Comment (brendank310) [inline common/serial/arch.h:13]: Agreed.

File: common/serial/Makefile (hunk 1 above)
Comment (rianquinn) [inline common/serial/Makefile:30]: Not sure yet.... I am about to run into the same problem once I do the VMCS work. We are probably going to have to define both, and have a unit test environment variable. For now I would leave this alone until we can address this (likely in a week)

File: common/serial/arch/arch_linux.h (hunk 6 above)
Comment (brendank310) [inline common/serial/arch/arch_linux.h]: Putting it there allowed for use in any consumer of arch_linux.h. Ultimately this macro goes away, as we'll have some proper debugging class.

File: common/serial/serial.h (hunk 9 above)
Comment (brendank310) [inline common/serial/serial.h]: Disagreed, perhaps the one liner one. Doesn't doxygen rely on C-style block comments?

File: common/serial/Makefile (hunk 3 above)
Comment (rianquinn) [inline common/serial/Makefile]: vpath is probably the simplest way to do it (really what vpath was designed for). So the code would be:

VPATH=arch
SOURCES = serial.cpp arch_linux.cpp

File: common/serial/arch/arch_linux.h (hunk 6 above)
Comment (rianquinn) [inline common/serial/arch/arch_linux.h]: agreed

File: common/serial/serial.h (hunk 9 above)
Comment (rianquinn) [inline common/serial/serial.h]: No... if you intend for doxygen to pickup on the definitions, use ///. I don't want to mix c and c++ comment styles. That is... for C use /\*  */ and for c++ use //

File: common/serial/port_io.h (hunk 12 above)
Comment (brendank310) [inline common/serial/port_io.h]: While I generally agree with the NASM approach, inline may be more appropriate in this scenario (one liners if done inline, probably no less than 10-20 if done separately.)

File: common/serial/serial.h (hunk 9 above)
Comment (rianquinn) [inline common/serial/serial.h]: This also applies for the doygen comments. For c its /*\* */ and for c++ its ///

File: common/serial/serial.h (hunk 13 above)
Comment (brendank310) [inline common/serial/serial.h]: While I'm generally against setters doing any error reporting, I think it's reasonable in these scenarios (the value is expected to belong to a small subset of the types domain).

File: common/serial/port_io.h (hunk 12 above)
Comment (rianquinn) [inline common/serial/port_io.h]: NASM vs GCC parameterized inline assembly will likely result in the same number of lines. The system V call convention is dirt simple, and you only use what you need. No stack crap, etc.... So the code would be something like outb RDI, RSI; ret 0

File: common/serial/serial.h (hunk 13 above)
Comment (rianquinn) [inline common/serial/serial.h]: That's really up to you. Either you use property evaluation for you unit tests, or state-based evaluation. It depends on how you want to report errors to the user. Qt kind of does it both ways because they are strict about encapsulation (which is stupid). Some classes have functions that report errors on setters, while others have an error() function that you have to call after you set. I prefer property based as your not having to store state in the class (never been a huge fan of errno like code)

File: common/serial/port_io.h (hunk 12 above)
Comment (rianquinn) [inline common/serial/port_io.h]: probably more like:

```
export port_write8

port_write8:
    outb rdi, rsi   ; rdi = first param, rsi = second param
    mov rax, 0    ; rax = return value
    ret
```

Comment (rianquinn) [inline common/serial/port_io.h]: I think that is a hell of a lot easier to read than GCC's ****ed up inline assembly, and once you add all of the parameters (which should be on a line of their own), your going to have basically the same thing.

File: common/serial/serial.cpp [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+    m_data_size = data_size;
+    m_stop_bits = stop_bits;
+}
+
+serial_port::~serial_port(void)
+{
+
+}
+
+int32_t
```
Comment (rianquinn) [inline common/serial/serial.cpp]: Instead of using int32_t, use an enum that is scoped to the class itself. For example:

```
class serial
{
public:

    enum error
    {
        success = 0,
        failure = 1
    }

public:

    serial() {}
    ~serial() {}

    serial::error do_something() { return success };
}
```

File: common/serial/port_io.h [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+public:
+	port_io(arch_handler *arch) : m_arch(arch) 
+	{
+
+	}
+
+	~port_io(void) 
+	{
+		delete m_arch;
```
Comment (rianquinn) [inline common/serial/port_io.h]: delete doesn't exist

File: common/serial/port_io.h (hunk 7 above)
Comment (rianquinn) [inline common/serial/port_io.h]: in general I would stay away from pointers. In C++, you can do subclassing without pointers just fine by passing everything by reference. The only time I would "point" to memory is when your talking to physical memory itself (i.e. an array, or to a mmio space, etc...). 

To get subclassing to work, just pass by reference. You will not be able to use pure virtual classes, so the base class will have to provide a default implementation (usually just ignore/error out), but the overrides will work just fine for the subclasses. 

In other words.... grep for "->" and find a way to change to "."

File: common/serial/portio_linux.h [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+extern uint8_t bf_inb(uint16_t port);
+extern uint16_t bf_inw(uint16_t port);
+
+#define TRACE() printf("%s:%d\n", __PRETTY_FUNCTION__, __LINE__)
+#define TRACE_INT(x) printf("[%s:%d] - %08X\n", __PRETTY_FUNCTION__, __LINE__, x)
+
+class portio_linux : public port_io {
+public:
+
+	portio_linux();
```
Comment (rianquinn) [inline common/serial/portio_linux.h:18]: the way I have been doing this is:

```
include \
    blah.h
    blah_base.h

src \
    arch \
        linux \
            blah.cpp
        vmm \
            blah.cpp
```

In the Makefile, you will add 
LINUX_SOURCES+= blah.cpp

The Cross-Compiled part has not been added to the makefiles yet (will do that soon), but it will likely end up with a 
VMM_SOURCES+=blah.cpp

Then add:
VPATH+=arch/linux arch/vmm

In your unit test, you will use blah_base in the hippomock.

Comment (rianquinn) [inline common/serial/portio_linux.h:18]: so your portio_linux would be just portio, and then your portio would turn into portio_base, which I think is more descriptive (and seems to be common practice)

File: common/serial/serial_port.h [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        SUCCESS,
+        GENERAL_ERROR,
+        INVALID_PARITY,
+        INVALID_BAUD,
+        INVALID_STOP_BITS,
+        INVALID_INTERRUPT_MODE,
+        UNIMPLEMENTED,
+    };
+
+    const char *toString[] = {
```
Comment (rianquinn) [inline common/serial/serial_port.h:18]: this only works if enum is 0->num of enums

File: common/serial/serial_port.h [hunk 18]
```diff
@@ -0,0 +1,104 @@
+#ifndef SERIAL_PORT__H
+#define SERIAL_PORT__H
+
+#define STRINGIFY(x) (#x)
+#include <bf_stdint.h>
+
+namespace serial {
+    enum errno {
+        SUCCESS,
```
Comment (rianquinn) [inline common/serial/serial_port.h:9]: I have ben using lowercase since these are not macros

File: common/serial/serial_port_x86.cpp [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+    // Update our internal storage of size
+    m_data_size = size;
+
+    data_bits = m_port_io.port_read_8(m_port + LCR_OFFSET);
+
+    while(data_size_table[i].data_size != 0)
+    {
+        if(size == data_size_table[i].data_size)
+        {
+            // Clear the data bits, leave the rest intact
```
Comment (rianquinn) [inline common/serial/serial_port_x86.cpp:211]: these comments can probably be removed

File: common/serial/serial_port_x86.h [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+ *  Provides a wrapper class around a serial port, which is
+ *  accessed via port I/O.
+ *  
+ *  Useful information:
+ *  http://wiki.osdev.org/Serial_Ports
+ *  http://www.sci.muni.cz/docs/pc/serport.txt
+ *
+ */
+
+#define COM1_IO_PORT 0x3f8
```
Comment (rianquinn) [inline common/serial/serial_port_x86.h:46]: use const instead of a macro, or enum

**Comment by rianquinn** [comment]
LGTM +1
